lib/identity_cache.rb

```diff
@@ -94,8 +94,9 @@ def should_fill_cache? # :nodoc:
     end
 
     def should_use_cache? # :nodoc:
-      pool = ActiveRecord::Base.connection_pool
-      !pool.active_connection? || pool.connection.open_transactions == 0
+      ActiveRecord::Base.connection_handler.connection_pool_list.none? do |pool|
+        pool.active_connection? && pool.connection.current_transaction.joinable?
```

Comment:
I added a (perhaps too long) comment.

The current implementation uses `TransactionManager#open_transactions`, which is also not part of the public Rails API - I do agree that `#joinable?` is a lot harder to understand, though.